blocks/bricks/__init__.py

```diff
@@ -499,35 +499,60 @@ class Softmax(Activation):
     def apply(self, input_):
         return tensor.nnet.softmax(input_)
 
-    @application
+    @application(inputs=['input_'], outputs=['output'])
+    def log_probabilities(self, input_):
+        """Normalize log-probabilities.
+
+        Converts unnormalized log-probabilities (exponents of which do not
+        sum to one) into actual log-probabilities (exponents of which sum
+        to one).
+
+        Parameters
+        ----------
+        input_ : :class:`~theano.Variable`
+            A matrix, each row contains unnormalized log-probabilities of a
+            distribution.
+
+        Returns
+        -------
+        output : :class:`~theano.Variable`
+            A matrix with normalized log-probabilities for each
+            distribution from `input_`.
+
+        """
+        shifted = input_ - input_.max(axis=1).dimshuffle(0, 'x')
+        return shifted - tensor.log(
+            tensor.exp(shifted).sum(axis=1).dimshuffle(0, 'x'))
+
+    @application(inputs=['y', 'x'], outputs=['output'])
     def categorical_cross_entropy(self, y, x):
-        """Return computationally stable softmax cost.
+        """Computationally stable cross-entropy for pre-softmax values.
 
         Parameters
         ----------
         y : :class:`~tensor.TensorVariable`
             In the case of a matrix argument, each slice along
-            axis represents one distribution. In the vector case, each
+            axis 0 represents one distribution. In the vector case, each
             element represents the position of the '1' in a one hot-vector.
         x : :class:`~tensor.TensorVariable`
-            Each slice along axis represents energies of a distribution,
-            that is pre-softmax values.
+            Each slice along axis 0 represents unnormalized
+            log-probabilities of a distribution, that is pre-softmax
+            values.
 
         Returns
         -------
         cost : :class:`~tensor.TensorVariable`
-            The cross entropy between y and x.
+            A vector of cross-entropies between distributions y and x.
 
         """
-        x = x - x.max(axis=1).dimshuffle(0, 'x')
-        log_prob = x - tensor.log(tensor.exp(x).sum(axis=1).dimshuffle(0, 'x'))
+        x = self.log_probabilities(x)
         if y.ndim == x.ndim - 1:
-            flat_log_prob = log_prob.flatten()
+            flat_log_prob = x.flatten()
```

Comment:
The one you point to also relies on non-GPU arange and advanced indexing.

Then I was recently profiling networks in Blocks and the performance 
killer is CAREduce which does e.g. gradient norm monitoring. It is very 
slow in Theano and simple monitoring can take several times more than 
fprob+bprop together.

Jan

Bart van Merriënboer wrote:
>
> In blocks/bricks/__init__.py
> <https://github.com/mila-udem/blocks/pull/759#discussion_r36050287>:
>
> > if y.ndim == x.ndim - 1:
> > - flat_log_prob = log_prob.flatten()
> > + flat_log_prob = x.flatten()
>
> Note that this requires advanced indexing, which was never supported
> on GPU (maybe it has been added lately?). This means that using
> |tensor.arange| like this forced the tensor to be transferred to CPU.
> The code was written the way it was to prevent that, see also e.g.
> https://github.com/lisa-lab/pylearn2/blob/master/pylearn2/models/mlp.py#L1422-L1440.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/blocks/pull/759/files#r36050287>.